include/ibamr/LevelSetUtilities.h

```diff
@@ -0,0 +1,420 @@
+     * the Heaviside function demarcating solid and fluid (fluid = liquid and gas) domains.
+     *
+     * Physically, these four integrals represents volume of the gas region, liquid region, solid region, and
+     * surface area of the fluid interface, respectively.
+     *
+     * \f$ \phi \f$ is taken to be positive in the liquid domain and negative in the gas domain.
+     *
+     * \f$ \Psi \f$ is taken to be positive outside the solid and negative inside the solid.
+     */
+    static std::vector<double> computeHeavisideIntegrals3PhaseFlows(LevelSetContainer* lsc);
+
+    /*!
+     * \return Integral of inflow \f$ -\vec{u} \cdot \vec{n}\f$ at a physical boundary.
+     *
+     * \param location_idx of the boundary at which the integral is evaluated
+     */
+    static double computeNetInflowPhysicalBoundary(SAMRAI::tbox::Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                                                   int u_idx,
+                                                   int bdry_loc_idx);
+
+    /*!
+     * \brief Class SetLSProperties is a utility class which sets (or resets after reinitialization)
+     * level set values on the patch hierarchy.
+     */
+    class SetLSProperties
+    {
+    public:
+        /*!
+         * The only constructor of this class.
+         */
+        SetLSProperties(const std::string& object_name, SAMRAI::tbox::Pointer<IBAMR::LSInitStrategy> ls_ops)
+            : d_object_name(object_name), d_ls_ops(ls_ops)
+        {
+            // intentionally left blank
+            return;
+        } // SetLSProperties
+
+        /*!
+         * Destructor for this class.
+         */
+        ~SetLSProperties() = default;
+
+        /*!
+         * Set the level set value on the patch hierarchy
+         */
+        void setLSData(int ls_idx,
+                       SAMRAI::tbox::Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                       const int integrator_step,
+                       const double current_time,
+                       const bool initial_time,
+                       const bool regrid_time);
+
+    private:
+        /*!
+         * Default constructor is not implemented and should not be used.
+         */
+        SetLSProperties() = delete;
+
+        /*!
+         * Default assignment operator is not implemented and should not be used.
+         */
+        SetLSProperties& operator=(const SetLSProperties& that) = delete;
+
+        /*!
+         * Default copy constructor is not implemented and should not be used.
+         */
+        SetLSProperties(const SetLSProperties& from) = delete;
+
+        /*!
+         * Name of this object.
+         */
+        std::string d_object_name;
+
+        SAMRAI::tbox::Pointer<IBAMR::LSInitStrategy> d_ls_ops;
+
+    }; // SetLSProperties
+
+    /*!
+     * \brief A function that sets or resets the level set data on the patch hierarchy
```

Comment:
What does this mean?

Reply:
This description is from Nishant. I think by sets he meant at t = 0, and by reset at other times. At initial time LS is set in a different manner than at other times.

Reply:
What data are being set / reset?

Reply:
Just the level set function (a scalar field).

Reply:
What is the algorithm that is used to set / reset?

Also, how does this relate to `SetLSProperties`, which at first glance appears to do the same thing (but with more flexibility since it takes in an `LSInitStrategy`)?

Reply:
Basically `setLSDataPatchHierarchy` is a callback function needed in `AdvDiffHI` to reset the level set function. This function (`setLSDataPatchHierarchy()`) takes a pointer to `SetLSProperties` to do the job. 

I think we had discussed many years ago to make `AdvDiffHI` agnostic of LS functions. This is done via callback functions to HI.